src/sidebar/components/button.js

```diff
@@ -0,0 +1,129 @@
+'use strict';
+
+const classnames = require('classnames');
+const propTypes = require('prop-types');
+const { createElement } = require('preact');
+
+const SvgIcon = require('./svg-icon');
+
+/**
+ * A button, one of three base types depending on provided props:
+ * - Icon-only button: `icon` present, `buttonText` missing
+ * - Text-only button: `buttonText` present, `icon` missing
+ * - Icon and text: both `icon` and `buttonText` present
+ *
+ * Buttons may be additionally styled with 0 to n of (none are mutually exclusive):
+ * - `useCompactStyle`: for fitting into tighter spaces
+ * - `useInputStyle`: for placing an icon-only button next to an input field
+ * - `usePrimaryStyle`: for applying "primary action" styling
+ * - `className`: arbitrary additional class name(s) to apply
+ */
+function Button({
+  buttonText = '',
+  className = '',
+  icon = '',
+  isActive = false,
+  onClick = () => null,
+  title,
+  useCompactStyle = false,
+  useInputStyle = false,
+  usePrimaryStyle = false,
+}) {
+  // If `buttonText` is provided, the `title` prop is optional and the `button`'s
+  // `title` attribute will be set from the `buttonText`
+  title = title || buttonText;
+
+  const baseClassName = buttonText ? 'button--labeled' : 'button--icon-only';
+
+  return (
+    <button
+      className={classnames(
+        'button',
+        baseClassName,
+        {
+          'button--compact': useCompactStyle,
+          'button--input': useInputStyle,
+          'button--primary': usePrimaryStyle,
+          'is-active': isActive,
+        },
+        className
+      )}
+      onClick={onClick}
+      aria-pressed={isActive}
+      title={title}
+    >
+      {icon && <SvgIcon name={icon} className="button__icon" />}
+      {buttonText}
+    </button>
+  );
+}
+
+/**
+ * Validation callback for `propTypes`. The given `propName` is conditionally
+ * required depending on the presence or absence of the `buttonText` property. Return
+ * an `Error` on validation failure, per propTypes API.
+ *
+ * @return {Error|undefined}
+ */
+function requiredStringIfButtonTextMissing(props, propName, componentName) {
+  if (
+    typeof props.buttonText !== 'string' &&
+    typeof props[propName] !== 'string'
+  ) {
+    return new Error(
+      `string property '${propName}' must be supplied to component '${componentName}'\
+if string property 'buttonText' omitted`
+    );
+  }
+  return undefined;
+}
+
+Button.propTypes = {
```

Comment:
Overall, no objections to this PR. I gave this some thought about how to make this more clearly show that there are 2 different types of modes this button can basically be in and I wanted to know what you thoughts about this sort of approach instead of writing a custom prop handler?

```
Button.propTypes = {

// type
icon: propTypes.shape({
    name: propTypes.string.isRequired,
    text: propTypes.string,
    title: propTypes.string.isRequired
}),

labeled: propTypes.shape({
    text: propTypes.string.isRequired,
    title: propTypes.string,
}),

// state
isActive: propTypes.bool,

// event
onClick: propTypes.func,

// styling 
className: propTypes.string,
useCompactStyle: propTypes.bool,
useInputStyle: propTypes.bool,
usePrimaryStyle: propTypes.bool,

}
```

That does not enforce one type or the other, and that could still be done with a custom wrapper similar to what you have done, but this is an attempt to make these props organized in a way that matches the component.